### ISSUE TITLE
STOW-RS not working

### ISSUE THREAD

**Issue by mohamadrezax** [DICOMcloud/DICOMcloud#9]
I'm trying test Dicomcloud on my local dev. enviroment , stow request responded with this XML:
`<?xml version="1.0" encoding="utf-8"?><NativeDicomModel>
  <DicomAttribute keyword="TransferSyntaxUID" tag="00020010" vr="UI">
    <Value number="1">1.2.840.10008.1.2.1</Value>
  </DicomAttribute>
  <DicomAttribute keyword="FailedSOPSequence" tag="00081198" vr="SQ">
    <Item number="1">
      <DicomAttribute keyword="SOPClassUID" tag="00080016" vr="UI">
        <Value number="1">1.2.840.10008.5.1.4.1.1.4</Value>
      </DicomAttribute>
      <DicomAttribute keyword="SOPInstanceUID" tag="00080018" vr="UI">
        <Value number="1">1.3.12.2.1107.5.2.30.27908.2017091907422330142800031</Value>
      </DicomAttribute>
      <DicomAttribute keyword="FailureReason" tag="00081197" vr="US">
        <Value number="1">272</Value>
      </DicomAttribute>
    </Item>
  </DicomAttribute>
  <DicomAttribute keyword="RetrieveURI" tag="0040E010" vr="UR">
    <Value number="1">https://localhost:44301/wadors/studies/</Value>
  </DicomAttribute>
</NativeDicomModel>`
according to http://dicom.nema.org/dicom/2013/output/chtml/part18/sect_6.6.html
the failure reason is : 0110 - Processing failure
The STOW-RS Service did not store the instance because of a general failure in processing the operation.
project is running without any errors and i can't find its reason.
can u help me?

**Comment by Zaid-Safadi** [comment]
Hi @mohamadrezax, since you have the server running locally. Can you run on debug and check what the exception is in the [WebObjectStoreService.cs](https://github.com/DICOMcloud/DICOMcloud/blob/development/DICOMcloud.Wado/Services/WebObjectStoreService.cs)  "StoreStudy" method?

**Comment by mohamadrezax** [comment]
Hi @zaid-safad, the exception says file name is too long. The problem can be solved with changing storage directory in web.config file. But i think the naming strategy in archive system can be enhanced to avoid such this problem

**Comment by Zaid-Safadi** [comment]
I agree @mohamadrezax that this is something that can be improved. 

The reason the file name is long is to produce a unique yet "FAST" way to retrieve the images from the server without doing a lookup. Also, when running this in the cloud "Azure" you really don't run into this issue for the file length limit.

One alternative that could be done is to hash the string (e.g. MD5 or SHA-1) to produce a shorter yet unique ID/file name. Let me know if you have any ideas.

**Comment by mohamadrezax** [comment]
This is good idea that application access files without any DB look-ups . For shortening strings that they are numbers you can using a simple conversion. Just convert them to hexadecimal or upper numerical basis @zaid-safadi

**Comment by Zaid-Safadi** [comment]
Seems there might be some options to improve this. I'll leave this issue open as an enhancement. 

Thanks for reporting @mohamadrezax

**Comment by Zaid-Safadi** [comment]
@mohamadrezax, I'm looking at the different options to shorten the name. Can you please give an example of how a long string/filename can be shortened with your suggested methods (hexadecimal or upper numerical basis)?

**Comment by Zaid-Safadi** [comment]
I have added a new class "HashedFileKeyProvider" that inherits from the former "FileKeyProvider" to provide an MD5 hashed key of the file path parts if they are longer than 32 characters. 

This new class is now used as the default when storing to a file system. 

When storing or reading and Media, the MediaId parts (transfer syntax, study uid, series uid, sop uid...) will be converted to their MD5 hash value and used for the file name.

I've tested this with +100 of images and they are working fine now. forming the path will have around 128 characters (32*4) for the instance ID which leaves plenty for the actual folder path (e.g. 
C:\Users\xxxx\AppData\Roaming\App_Data\Storage\ds\3EDC6E2C9F9230CA29E55054F6FA1ECA\0F275DD37FC7DC58F04FC0C95B980055\0D65842681EEF19C19E0106F0352DC52\8E3A044CFEFB4EDC3982EB11387A7130
